### PR TITLE
Add scripts to update server & apps or apps only

### PR DESCRIPTION
**Pull request by jancborchardt**
Ref https://github.com/nextcloud/server/issues/15145#issuecomment-490570313

One script `./build/update.sh` to simply update server and all apps to current master.
And another one `./build/update-apps.sh` to only update the apps.

Please check @MorrisJobke @ChristophWurst @nickvergessen @skjnldsv 

Possible future enhancements:
- The branch could be supplied by an optional argument like `./build/update.sh stable14` but many apps don’t seem to really use these branches.
- If that’s wanted we could have another script for initial setup, including also cloning the repositories of the apps which are shipped, and optionally also "core" apps like Calendar, Contacts, Talk, Mail, etc.

### CODE REVIEW

**Comment by jancborchardt** [comment]
Also @camilasan check if you like, this basically replaces the update scripts from my repository https://github.com/jancborchardt/nextcloud-scripts/blob/master/nextcloud-update.sh :)

**Comment by MorrisJobke** [comment]
> * The branch could be supplied by an optional argument like `./build/update.sh stable14` but many apps don’t seem to really use these branches.

I actually have that here as well. For all apps that don't have this it will stay just in the branch they are.

```
find apps* -name .git -exec git --git-dir={} --work-tree={}/.. checkout master \;
find apps* -name .git -exec git --git-dir={} --work-tree={}/.. checkout stable16 \;
```

As this is quite fast I execute the update script right afterwards and then it's good to go.

File: build/update.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+# Update Nextcloud server and apps from latest git master
+# For local development environment
+# Use from Nextcloud server folder with `./build/update.sh`
+
+# Update server
+printf "\n\033[1m${PWD##*/}\033[0m\n"
+git checkout master
+git pull --quiet
```
Comment (MorrisJobke) [inline build/update.sh]: Could we make it like this to also remove remote branches that are deleted on GitHub?

```suggestion
git pull --quiet -p
```

File: build/update-apps.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+#
+# It automatically:
+# - goes through all apps which are not shipped via server
+# - shows the app name in bold and uses whitespace for separation
+# - changes to master and pulls quietly
+# - shows the 3 most recent commits for context
+# - removes branches merged into master
+# - … could even do the build steps if they are consistent for the apps (like `make`)
+
+find apps* -maxdepth 2 -name .git -exec sh -c 'cd {}/../ && printf "\n\033[1m${PWD##*/}\033[0m\n" && git checkout master && git pull --quiet && git --no-pager log -3 --pretty=format:"%h %Cblue%ar%x09%an %Creset%s" && printf "\n" && git branch --merged master | grep -v "master$" | xargs --no-run-if-empty git branch -d && git fetch --prune --quiet && cd ..' \;
```
Comment (MorrisJobke) [inline build/update-apps.sh]: Same here:

```suggestion
find apps* -maxdepth 2 -name .git -exec sh -c 'cd {}/../ && printf "\n\033[1m${PWD##*/}\033[0m\n" && git checkout master && git pull --quiet -p && git --no-pager log -3 --pretty=format:"%h %Cblue%ar%x09%an %Creset%s" && printf "\n" && git branch --merged master | grep -v "master$" | xargs --no-run-if-empty git branch -d && git fetch --prune --quiet && cd ..' \;
```

File: build/update.sh (hunk 1 above)
Comment (jancborchardt) [reply build/update.sh]: Isn’t that done by the `git fetch --prune --quiet` further down?

File: build/update-apps.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Comment (kesselb) [inline build/update-apps.sh]:
```suggestion
#!/usr/bin/env bash
```
https://stackoverflow.com/questions/16365130/what-is-the-difference-between-usr-bin-env-bash-and-usr-bin-bash

File: build/update.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (kesselb) [inline build/update.sh]:
```suggestion
#!/usr/bin/env bash
```

File: build/update.sh (hunk 1 above)
Comment (MorrisJobke) [reply build/update.sh]: > Isn’t that done by the `git fetch --prune --quiet` further down?

Yes it is, but you could do this in one go, by doing it in the `pull`

File: build/update-apps.sh (hunk 3 above)
Comment (jancborchardt) [reply build/update-apps.sh]: @kesselb So according to that post, there are benefits but also drawbacks (e.g. if we ever want to pass arguments like branches etc). Does this change actually fix an issue, like not working on a specific OS?

**Comment by jancborchardt** [comment]
Pulled in @MorrisJobke’s suggestions, but not sure about @kesselb’s (see reply), any input?

File: build/update-apps.sh (hunk 3 above)
Comment (kesselb) [reply build/update-apps.sh]: > Does this change actually fix an issue, like not working on a specific OS?

No ;) Go ahead with `#!/bin/bash`. If someone complains we can change it later.

Comment (jancborchardt) [reply build/update-apps.sh]: Hehe :) Ok, will resolve this and the other comment then. Does the rest look good? If so feel free to add your approval review.

**Review by kesselb — APPROVED** [review]
Works for me (Ubuntu 19.04)

File: build/update-apps.sh [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+#
+# It automatically:
+# - goes through all apps which are not shipped via server
+# - shows the app name in bold and uses whitespace for separation
+# - changes to master and pulls quietly
+# - shows the 3 most recent commits for context
+# - removes branches merged into master
+# - … could even do the build steps if they are consistent for the apps (like `make`)
+
+find apps* -maxdepth 2 -name .git -exec sh -c 'cd {}/../ && printf "\n\033[1m${PWD##*/}\033[0m\n" && git checkout master && git pull --quiet -p && git --no-pager log -3 --pretty=format:"%h %Cblue%ar%x09%an %Creset%s" && printf "\n" && git branch --merged master | grep -v "master$" | xargs --no-run-if-empty git branch -d && git fetch --prune --quiet && cd ..' \;
```
Comment (MorrisJobke) [inline build/update-apps.sh]: Short question, why do you run `git pull --quiet -p` and afterwards a `git fetch --prune --quiet`? The latter shouldn't be needed because the `git pull` is the shorthand for `git fetch` followed by a `git merge FETCH_HEAD`. This would then dramatically speed up the whole command, because only one fetch is needed (which takes most of the time on my side due to the network IO).

File: build/update.sh [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+# Use from Nextcloud server folder with `./build/update.sh`
+
+# Update server
+printf "\n\033[1m${PWD##*/}\033[0m\n"
+git checkout master
+git pull --quiet -p
+git --no-pager log -3 --pretty=format:"%h %Cblue%ar%x09%an %Creset%s"
+printf "\n"
+git branch --merged master | grep -v "master$" | xargs --no-run-if-empty git branch -d
+git fetch --prune --quiet
```
Comment (MorrisJobke) [inline build/update.sh]: Same here.

File: build/update-apps.sh (hunk 5 above)
Comment (jancborchardt) [reply build/update-apps.sh]: Sorry – I applied your code change suggestion and then forgot to make the additional change. Fixed now.

File: build/update.sh (hunk 6 above)
Comment (jancborchardt) [reply build/update.sh]: Fixed now too :)

File: build/update-apps.sh [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+#
+# It automatically:
+# - goes through all apps which are not shipped via server
+# - shows the app name in bold and uses whitespace for separation
+# - changes to master and pulls quietly
+# - shows the 3 most recent commits for context
+# - removes branches merged into master
+# - … could even do the build steps if they are consistent for the apps (like `make`)
+
+find apps* -maxdepth 2 -name .git -exec sh -c 'cd {}/../ && printf "\n\033[1m${PWD##*/}\033[0m\n" && git checkout master && git pull --quiet -p && git --no-pager log -3 --pretty=format:"%h %Cblue%ar%x09%an %Creset%s" && printf "\n" && git branch --merged master | grep -v "master$" | xargs --no-run-if-empty git branch -d && cd ..' \;
```
Comment (MorrisJobke) [inline build/update-apps.sh]: The only problem on any non-GNU system (*BSD, macOS) is that `--no-run-if-empty` is not available there. Is this really needed, because locally it works fine without this option.

Comment (jancborchardt) [reply build/update-apps.sh]: I added it to resolve the error "fatal: branch name required" on the `git branch -d` when there are no branches which are not merged.

Does having this option make it crash for you, or just brings up an error?

I found this possible solution which is said to run on both Linux and macOS: https://stackoverflow.com/a/17411766

I adjusted it for ours, but then it gives me "error: branch ':' not found."
```
git branch --merged master | (grep -v "master$" || echo :) | xargs git branch -d
```


**TL;DR: If it makes the command not work on macOS, let’s remove the flag, as it only causes an error on Linux but the command still works.**

Comment (MorrisJobke) [reply build/update-apps.sh]: > TL;DR: If it makes the command not work on macOS, let’s remove the flag, as it only causes an error on Linux but the command still works.

yes - it prints an error, but does not execute the command

Comment (jancborchardt) [reply build/update-apps.sh]: Fixed now, removed the extra flag. :)

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by jancborchardt** [comment]
Yay, thanks @MorrisJobke @kesselb! :) Am looking forward to slowly getting rid of my custom scripts repo and transferring the stuff to either this repo or the docs to be more universally usable. :tada:
